### PR TITLE
mzworkflows: allow Python workflows to run Steps

### PR DESCRIPTION
**Pull request by philip-stoev**
Here is what I came up in terms of Python code to be able to run Steps. I decided to leverage your `@Step` annotation so that there is no need to add any additional code as more steps are added. As per your original idea I also install each step as a method under the same name, however as a classmethod to the `Steps` rather than `Composition` . Let me know if you agree.

The syntax of the test now becomes:

```
from materialize.mzcompose import Steps as S
from materialize.mzcompose import Workflow

def workflow_in_python(w: Workflow):
    S.start_services(w, services=["materialized"])
    S.wait_for_mz(w, service="materialized")
    S.kill_services(w, services=["materialized"])
```

Which is reasonably concise. What do you think? Thanks.

### CODE REVIEW

File: misc/python/materialize/cli/mzcompose.py [hunk 1]
```diff
@@ -18,9 +18,8 @@
 from pathlib import Path
 from typing import IO, List, Optional, Sequence, Text, Tuple
 
-from typing_extensions import NoReturn
-
 from materialize import errors, mzbuild, mzcompose, spawn, ui
+from typing_extensions import NoReturn
```
Comment (philip-stoev) [inline misc/python/materialize/cli/mzcompose.py:22]: Ops, I pushed an import order change in an unrelated file. What is the deal with those? I keep getting them every time I run bin/pylint

File: misc/python/materialize/mzcompose.py [hunk 2]
```diff
@@ -197,7 +196,7 @@ def __init__(self, repo: mzbuild.Repository, name: str):
             compose = yaml.safe_load(f)
 
         # Stash away sub workflows so that we can load them with the correct environment variables
-        self.yaml_workflows = compose.pop("mzworkflows", None)
+        self.yaml_workflows = compose.pop("mzworkflows", None) or {}
```
Comment (philip-stoev) [inline misc/python/materialize/mzcompose.py:199]: This allows for a `mzcompose.yml` file wilth no `workflows` section.

Comment (quodlibetor) [reply misc/python/materialize/mzcompose.py:199]: This can be more concisely expressed as:

```suggestion
        self.yaml_workflows = compose.pop("mzworkflows", {}})
```

File: misc/python/materialize/mzcompose.py [hunk 3]
```diff
@@ -619,18 +618,34 @@ def named(cls, name: str) -> Type["WorkflowStep"]:
...
     @classmethod
-    def register(cls, name: str) -> Callable[[Type[T]], Type[T]]:
+    def register(
+        cls, name: str
+    ) -> Callable[[Type["WorkflowStep"]], Type["WorkflowStep"]]:
         if name in cls._steps:
             raise ValueError(f"Double registration of step name: {name}")
 
-        def reg(to_register: Type[T]) -> Type[T]:
+        def reg(to_register: Type["WorkflowStep"]) -> Type["WorkflowStep"]:
```
Comment (quodlibetor) [inline misc/python/materialize/mzcompose.py:627]: Was this causing problems? This type renaming is slightly worse in the abstract: it may make mypy believe that you have erased the type of the step that you have registered. The `T` should tell mypy that `register(obj)` returns exactly the same type of object that you passed to it.

File: misc/python/materialize/mzcompose.py [hunk 4]
```diff
@@ -619,18 +618,34 @@ def named(cls, name: str) -> Type["WorkflowStep"]:
...
+                step: WorkflowStep = to_register(**args)
+                step.run(workflow)
+
+            func_name = name.replace("-", "_")
+            if not hasattr(cls, func_name):
+                setattr(cls, func_name, run_step)
+            else:
+                raise errors.Failed(
+                    f"Unable to register method Step.{func_name} as one already exists."
+                )
```
Comment (quodlibetor) [inline misc/python/materialize/mzcompose.py:646]: I think it would be easier to use this (autocompletion would work, mypy wouldn't complain) if we just made `run_step()` a global function in this module and required every step to define themselves as a function:

```python
# unfortunately handling higher order functions with kwargs is rough
# https://stackoverflow.com/a/61583234/25616
def runnable(step: WorkflowStep) -> Callable[[Workflow, Any], None]:
    def run_step(workflow: Workflow, **args: Dict) -> None:
       step: WorkflowStep = step(**args)
       step.run(workflow)
    run_step.name = step.name
    return run_step


@Steps.register("good-name")
class GoodNameStep(WorkflowStep): ...

good_name = runnable(GoodNameStep)
```

I would really like for us to make WorkflowStep classes unnecessary entirely and just create a decorator that can work on functions directly. I haven't gotten around to it yet, but it would be great.

File: misc/python/materialize/cli/mzcompose.py (hunk 1 above)
Comment (quodlibetor) [reply misc/python/materialize/cli/mzcompose.py:22]: I believe that this might be the `isort` call in `bin/pyfmt`

**Review by quodlibetor — COMMENTED** [review]
I love what this will allow us to do! I have some concerns about the usability of this from client code with some possible solutions in a comment.

File: misc/python/materialize/mzcompose.py (hunk 3 above)
Comment (philip-stoev) [reply misc/python/materialize/mzcompose.py:627]: `mypy` was complaining that `T` is not guaranteed to have a `run()` method that was being used below,

File: misc/python/materialize/mzcompose.py (hunk 4 above)
Comment (philip-stoev) [reply misc/python/materialize/mzcompose.py:646]: can I auto-register the steps as global functions . This will not work with autocompletion, but will not require every new step to be registered separately as a global function in the future. 

I think the Step classes should stay as you can then ask them to pretty-print themselves, dump themselves back to YAML and other such stuff, if ever necessary.

Comment (benesch) [reply misc/python/materialize/mzcompose.py:646]: They shouldn't be global functions but just methods on the `Workflow` class. That would be most ergonomic, I think. 

Long term I would like to entirely deprecate writing workflows in YAML.

**Comment by benesch** [comment]
This is great, @philip-stoev! Thanks very much for doing this. I do think that the approach you've taken here of dynamically constructing the functions at @Step.register time is correct for the short term. As I mentioned above, long term I think we should remove the ability to construct YAML workflows, and then the step functions can be simple methods on Workflow.

File: misc/python/materialize/mzcompose.py (hunk 3 above)
Comment (benesch) [reply misc/python/materialize/mzcompose.py:627]: This doesn't sound right to me. mypy will now statically prevent you from calling `Step.register` on something that is not a `WorkflowStep`, which is correct. It also means we can remove the explicit `issubclass` check inside the function, since the type check is now enforced statically.

**Comment by benesch** [comment]
Concretely, here is the diff I recommend to make these methods on `Workflow` rather than class methods on `Step`. LGTM with this patch.

```diff
diff --git a/misc/python/materialize/mzcompose.py b/misc/python/materialize/mzcompose.py
index 8417b9880..1f1c67a15 100644
--- a/misc/python/materialize/mzcompose.py
+++ b/misc/python/materialize/mzcompose.py
@@ -196,7 +196,7 @@ class Composition:
             compose = yaml.safe_load(f)
 
         # Stash away sub workflows so that we can load them with the correct environment variables
-        self.yaml_workflows = compose.pop("mzworkflows", None) or {}
+        self.yaml_workflows = compose.pop("mzworkflows", {})
 
         # Load the mzworkflows.py file, if any.
         mzworkflows_py = self.path.parent / "mzworkflows.py"
@@ -625,24 +625,24 @@ class Steps:
             raise ValueError(f"Double registration of step name: {name}")
 
         def reg(to_register: Type["WorkflowStep"]) -> Type["WorkflowStep"]:
-            if not issubclass(to_register, WorkflowStep):
-                raise ValueError(
-                    f"Registered step must be a WorkflowStep: {to_register}"
-                )
             cls._steps[name] = to_register
             to_register.name = name
 
-            # Allow the step to also be called as a Step.step_name() classmethod
-            def run_step(workflow: Workflow, **args: Dict) -> None:
-                step: WorkflowStep = to_register(**args)
+            # Allow the step to also be called as a Workflow.step_name() classmethod
+            def run_step(workflow: Workflow, **kwargs: Any) -> None:
+                step: WorkflowStep = to_register(**kwargs)
                 step.run(workflow)
 
             func_name = name.replace("-", "_")
-            if not hasattr(cls, func_name):
-                setattr(cls, func_name, run_step)
+            if func_name == "run":
+                # Temporary workaround for the fact that `Workflow.run` already
+                # exists.
+                func_name = "run_service"
+            if not hasattr(Workflow, func_name):
+                setattr(Workflow, func_name, run_step)
             else:
                 raise errors.Failed(
-                    f"Unable to register method Step.{func_name} as one already exists."
+                    f"Unable to register method Workflow.{func_name} as one already exists."
                 )
 
             return to_register
diff --git a/test/mzworkflows_py/mzworkflows.py b/test/mzworkflows_py/mzworkflows.py
index 5b30948e4..40d21e6cb 100644
--- a/test/mzworkflows_py/mzworkflows.py
+++ b/test/mzworkflows_py/mzworkflows.py
@@ -7,11 +7,10 @@
 # the Business Source License, use of this software will be governed
 # by the Apache License, Version 2.0.
 
-from materialize.mzcompose import Steps as S
 from materialize.mzcompose import Workflow
 
 
 def workflow_in_python(w: Workflow):
-    S.start_services(w, services=["materialized"])
-    S.wait_for_mz(w, service="materialized")
-    S.kill_services(w, services=["materialized"])
+    w.start_services(services=["materialized"])
+    w.wait_for_mz(service="materialized")
+    w.kill_services(services=["materialized"])
```

File: misc/python/materialize/mzcompose.py (hunk 2 above)
Comment (philip-stoev) [reply misc/python/materialize/mzcompose.py:199]: Thanks, done.

**Comment by philip-stoev** [comment]
@benesch thanks for the patch, I was just coding it myself (sans the `Temporary workaround for the fact that `Workflow.run` already` part) when I saw yours .
